tensorly/decomposition/_nn_cp.py

```diff
@@ -0,0 +1,413 @@
+    r"""Initialize factors used in `parafac`.
+
+    The type of initialization is set using `init`. If `init == 'random'` then
+    initialize factor matrices using `random_state`. If `init == 'svd'` then
+    initialize the `m`th factor matrix using the `rank` left singular vectors
+    of the `m`th unfolding of the input tensor.
+
+    Parameters
+    ----------
+    tensor : ndarray
+    rank : int
+    init : {'svd', 'random'}, optional
+    svd : str, default is 'numpy_svd'
+        function to use to compute the SVD, acceptable values in tensorly.SVD_FUNS
+    nntype : {'nndsvd', 'nndsvda'}
+        Whether to fill small values with 0.0 (nndsvd), or the tensor mean (nndsvda, default).
+
+    Returns
+    -------
+    factors : CPTensor
+        An initial cp tensor.
+
+    """
+    rng = check_random_state(random_state)
+
+    if init == 'random':
+        kt = random_cp(tl.shape(tensor), rank, normalise_factors=False, **tl.context(tensor))
+
+    elif init == 'svd':
+        try:
+            svd_fun = tl.SVD_FUNS[svd]
+        except KeyError:
+            message = 'Got svd={}. However, for the current backend ({}), the possible choices are {}'.format(
+                    svd, tl.get_backend(), tl.SVD_FUNS)
+            raise ValueError(message)
+
+        factors = []
+        for mode in range(tl.ndim(tensor)):
+            U, S, V = svd_fun(unfold(tensor, mode), n_eigenvecs=rank)
+
+            # NNDSVD initialization
+            W = tl.zeros_like(U)
+            H = tl.zeros_like(V)
+
+            # The leading singular triplet is non-negative
+            # so it can be used as is for initialization.
+            W = tl.index_update(W, tl.index[:, 0], tl.sqrt(S[0]) * tl.abs(U[:, 0]))
+            H = tl.index_update(H, tl.index[0, :], tl.sqrt(S[0]) * tl.abs(V[0, :]))
+
+            for j in range(1, tl.shape(U)[1]):
+                x, y = U[:, j], V[j, :]
+
+                # extract positive and negative parts of column vectors
+                x_p, y_p = tl.clip(x, a_min=0.0), tl.clip(y, a_min=0.0)
+                x_n, y_n = tl.abs(tl.clip(x, a_max=0.0)), tl.abs(tl.clip(y, a_max=0.0))
+
+                # and their norms
+                x_p_nrm, y_p_nrm = tl.norm(x_p), tl.norm(y_p)
+                x_n_nrm, y_n_nrm = tl.norm(x_n), tl.norm(y_n)
+
+                m_p, m_n = x_p_nrm * y_p_nrm, x_n_nrm * y_n_nrm
+
+                # choose update
+                if m_p > m_n:
+                    u = x_p / x_p_nrm
+                    v = y_p / y_p_nrm
+                    sigma = m_p
+                else:
+                    u = x_n / x_n_nrm
+                    v = y_n / y_n_nrm
+                    sigma = m_n
+
+                lbd = tl.sqrt(S[j] * sigma)
+                W = tl.index_update(W, tl.index[:, j], lbd * u)
+                H = tl.index_update(H, tl.index[j, :], lbd * v)
+
+            # After this point we no longer need H
+            idx = W < np.finfo(np.float32).eps
```

Comment:
This could be just `eps = tl.eps(tensor.dtype)` and then using it as in the comments below.